streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestorer.java

```diff
@@ -51,14 +58,34 @@ long checkpoint() {
         return checkpoint == null ? NO_CHECKPOINT : checkpoint;
     }
 
-    void restore(final byte[] key, final byte[] value) {
-        stateRestoreCallback.restore(key, value);
+    void maybeNotifyRestoreStarted(long startingOffset, long endingOffset) {
+        if (!restoreStarted) {
+            compositeRestoreListener.onRestoreStart(partition, storeName, startingOffset, endingOffset);
+            restoreStarted = true;
+        }
+    }
+
+    void restoreDone() {
+        compositeRestoreListener.onRestoreEnd(partition, storeName, restoredNumRecords());
+        restoreStarted = false;
+    }
+
+    void restoreBatchCompleted(long currentRestoredOffset, int numRestored) {
```

Comment:
These two functions can be merged into one?

```
restore(..) {
    listener.restoreAll();
    listener.onBatchRestored();
}
```

See my other comments on the `StoreChangelogReader.java` class.